compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1236,6 +1236,37 @@ impl<'a> Parser<'a> {
 
     /// Parse an indexing expression `expr[...]`.
     fn parse_index_expr(&mut self, lo: Span, base: P<Expr>) -> PResult<'a, P<Expr>> {
+        if self.prev_token.is_ident()
+            && matches!(self.token.kind, token::OpenDelim(Delimiter::Bracket))
+            && self.look_ahead(1, |t| t.is_lit())
+            && self.look_ahead(2, |t| matches!(t.kind, token::Semi))
+            && self.look_ahead(3, |t| t.is_lit())
+            && self.look_ahead(4, |t| matches!(t.kind, token::CloseDelim(Delimiter::Bracket)))
```

Comment:
This could probably be done with a [parser snapshot](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/parser/struct.Parser.html#method.create_snapshot_for_diagnostic) instead of so much lookahead. You should look around the parser code to see how it's used if you're unfamiliar with it.

Reply:
Actually, nevermind. We need this lookahead because we've not yet hit an error. :/